src/webservice/Router.cpp

```diff
@@ -0,0 +1,128 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "webservice/Router.h"
+
+#include <sstream>
+
+#include <proxygen/httpserver/RequestHandler.h>
+
+#include "webservice/NotFoundHandler.h"
+
+namespace nebula {
+
+namespace web {
+
+void Route::checkPath(const std::string &path) {
+    CHECK(!path.empty() && path[0] == '/') << "Path must start with '/'";
+}
+
+void Route::setPath(const std::string &path) {
+    auto loc = path.find(':');
+    if (loc != std::string::npos) {
+        createPattenRegex(path);
+        this->path_ = path;
+    } else {
+        // exact path, remove the last / of path
+        if (path.size() > 1 && path.back() == '/') {
+            this->path_ = path.substr(0, path.length() - 1);
+        } else {
+            this->path_ = path.empty() ? "/" : path;
+        }
+    }
+}
+
+// static
+std::unique_ptr<std::regex> Route::reToken_ =
+    std::make_unique<std::regex>(":([A-Za-z][A-Za-z_0-9]*)");
+
+void Route::createPattenRegex(const std::string &path) {
+    int pos = 0;
+    std::stringstream ss;
+    for (std::sregex_iterator next(path.begin(), path.end(), *reToken_), end; next != end; next++) {
+        std::string str = next->str();
+        if (!str.empty() && str.front() == ':') {
+            str = str.substr(1);
+        }
+        for (auto &group : groups_) {
+            if (str == group) {
+                LOG(FATAL) << "Cannot use identifier " << group
+                           << " more than once in pattern string";
+            }
+        }
+        groups_.emplace_back(std::move(str));
+        ss << path.substr(pos, next->position() - pos) << "([^/]+)";
+        pos = next->position() + next->str().length();
+    }
+    ss << path.substr(pos);
+    this->pattern_ = std::make_unique<std::regex>(ss.str());
+}
+
+bool Route::matches(proxygen::HTTPMethod method, const std::string &path) const {
+    if (method_ != method) {
+        return false;
+    }
+    std::string p = path;
+    if (p.empty()) {
+        p = "/";
+    } else if (p.size() > 1 && p.back() == '/') {
+        p.pop_back();
+    }
+    if (pattern_) {
+        return std::regex_search(p, *pattern_);
+    } else {
+        return path_ == p;
+    }
+}
+
+Router::~Router() {
+    while (head_ != nullptr) {
+        auto next = head_->next();
+        delete head_;
+        head_ = next;
+    }
+}
+
+Route &Route::handler(ReqHandlerGenerator generator) {
+    CHECK(generators_.empty()) << "Only allowed to register handler once for a route";
+    generators_.push_back(generator);
```

Comment:
If generators_ has only one element,  why use vector?

Reply:
In the beginning, I want to allow to register multiple handlers for one route path. But now I fix it for better code reading.